feat/upgrade binance spot connector

aarmoa opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
PR to update Binance spot connector including all new tools and functionalities that are considered standard for connectors.
- Binance in flight order has been updated.
- Order book data source class has been updated.
- User stream data source class has been updated.


**Tests performed by the developer**:
All unit tests passing in green.


**Tips for QA testing**:
In this moment the only thing to test is the functionality related to the administration of the order book.

[ch21083]

PtrckM commented:
- getting error when using the command `open_orders` or `open_orders --full_report`
  - this doesn't happen on latest release and dev branch 

![image](https://user-images.githubusercontent.com/54516858/147319628-10721c29-aaa5-400e-8422-5039f65b02fb.png)

```
2021-12-24 03:05:59,075 - 13447 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'BCCBTC'
Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/cachetools/cache.py", line 39, in __getitem__
    return self.__data[key]
KeyError: '((), {})'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/upgrade_binance_spot_connector/hummingbot/core/utils/__init__.py", line 15, in memoize
    return cache[key]
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/cachetools/ttl.py", line 84, in __getitem__
    return cache_getitem(self, key)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/cachetools/cache.py", line 41, in __getitem__
    return self.__missing__(key)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/cachetools/cache.py", line 68, in __missing__
    raise KeyError(key)
KeyError: '((), {})'

During handling of the above exception, another exception occurred:
```
- logs, configs and data

[binance.zip](https://github.com/CoinAlpha/hummingbot/files/7772758/binance.zip)

rxlxrxsx commented:
- Confirming that the issue reported above is now fixed on latest commit.

- Getting `Error parsing the trading pair rule` immediately after starting the bot on the latest commit.

<img width="1426" alt="Screen Shot 2021-12-26 at 12 44 46 PM" src="https://user-images.githubusercontent.com/64377055/147399998-8faa8877-41b8-4067-9ba2-0a709361a75b.png">

Sample error:

```python
2021-12-26 13:21:44,015 - 10400 - hummingbot.connector.exchange.binance.binance_exchange - ERROR - Error parsing the trading pair rule {'symbol': 'HSRBTC', 'status': 'BREAK', 'baseAsset': 'HSR', 'baseAssetPrecision': 8, 'quoteAsset': 'BTC', 'quotePrecision': 8, 'quoteAssetPrecision': 8, 'baseCommissionPrecision': 8, 'quoteCommissionPrecision': 8, 'orderTypes': ['LIMIT', 'LIMIT_MAKER', 'MARKET', 'STOP_LOSS_LIMIT', 'TAKE_PROFIT_LIMIT'], 'icebergAllowed': True, 'ocoAllowed': True, 'quoteOrderQtyMarketAllowed': True, 'isSpotTradingAllowed': True, 'isMarginTradingAllowed': False, 'filters': [{'filterType': 'PRICE_FILTER', 'minPrice': '0.00000100', 'maxPrice': '1000.00000000', 'tickSize': '0.00000100'}, {'filterType': 'PERCENT_PRICE', 'multiplierUp': '5', 'multiplierDown': '0.2', 'avgPriceMins': 5}, {'filterType': 'LOT_SIZE', 'minQty': '0.01000000', 'maxQty': '92141578.00000000', 'stepSize': '0.01000000'}, {'filterType': 'MIN_NOTIONAL', 'minNotional': '0.00010000', 'applyToMarket': True, 'avgPriceMins': 5}, {'filterType': 'ICEBERG_PARTS', 'limit': 10}, {'filterType': 'MAX_NUM_ORDERS', 'maxNumOrders': 200}, {'filterType': 'MAX_NUM_ALGO_ORDERS', 'maxNumAlgoOrders': 5}], 'permissions': ['SPOT']}. Skipping.
Traceback (most recent call last):
  File "hummingbot/connector/exchange/binance/binance_exchange.pyx", line 367, in hummingbot.connector.exchange.binance.binance_exchange.BinanceExchange._format_trading_rules
    trading_pair = await BinanceAPIOrderBookDataSource.trading_pair_associated_to_exchange_symbol(
  File "/Users/reiniersalazar/hummingbot_c1/hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py", line 155, in trading_pair_associated_to_exchange_symbol
    return symbol_map[symbol]
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/bidict/_base.py", line 387, in __getitem__
    return self._fwdm[key]
KeyError: 'HSRBTC'
```

Logs and config: [binance.zip](https://github.com/CoinAlpha/hummingbot/files/7776600/binance.zip)

RC-13 commented:
Adding additional issue seen while checking the PR:
- When checking for balance and the rate_oracle_source has defaulted to Binance, the following error keeps coming up:
`2021-12-27 00:03:07,899 - 31540 - hummingbot.core.rate_oracle.rate_oracle - ERROR - Unexpected error while retrieving rates from Coingecko. Check the log file for more info.`
![image](https://user-images.githubusercontent.com/57189990/147424423-bfbeac96-795d-4760-bbdd-028d99f4e626.png)
This error doesn't persist if the rate_oracle is changed to Ascend_Ex, Coingecko, or KuCoin.

aarmoa commented:
> Adding additional issue seen while checking the PR:
> 
> * When checking for balance and the rate_oracle_source has defaulted to Binance, the following error keeps coming up:
>   `2021-12-27 00:03:07,899 - 31540 - hummingbot.core.rate_oracle.rate_oracle - ERROR - Unexpected error while retrieving rates from Coingecko. Check the log file for more info.`
>   ![image](https://user-images.githubusercontent.com/57189990/147424423-bfbeac96-795d-4760-bbdd-028d99f4e626.png)
>   This error doesn't persist if the rate_oracle is changed to Ascend_Ex, Coingecko, or KuCoin.

@RC-13 Do you still have handy the logs for this error?

rapcmia commented:
Hi @aarmoa 
I tried to reproduce it and it does not seem to show any logs despite the error message telling check the log file.

![Screenshot 2021-12-28 212718](https://user-images.githubusercontent.com/73840223/147572644-065a9ae5-67be-445a-8509-a0fd7167bd94.png)

```
2021-12-28 21:25:48,637 - 4472 - hummingbot.core.rate_oracle.rate_oracle - ERROR - Unexpected error while retrieving rates from Coingecko. Check the log file for more info.
2021-12-28 21:26:25,823 - 4472 - hummingbot.core.rate_oracle.rate_oracle - ERROR - Unexpected error while retrieving rates from Coingecko. Check the log file for more info.
```

I also tried to:
- import a strategy then start and check the balance, the error does not seem to occur but the balance issue still persist
- change the `global token` from `USD` to `USDT` or vice versa for Binance, getting the same error 

Please see attached for the log files:
[pmm-test.zip](https://github.com/CoinAlpha/hummingbot/files/7784976/pmm-test.zip)

aarmoa commented:
I think I was able to solve the issue with the RateOracle.
I have also committed the changes that convert the BinanceExchange class to pure Python, and it is no longer using the third party library to communicate with the server.
I could not test the connectivity properly because or problems with the testing accounts. I will discuss the details with QA team, but in the meantime I will need them to test the orders related operations are working properly.

rxlxrxsx commented:
> I think I was able to solve the issue with the RateOracle.
- Confirmed that the issue rate oracle has been fixed on the latest commit.

- On the latest commit as well, I'm getting an API error from Binance that prevents orders from being created. I have cross checked this on the master and development branch but haven't got the same issue.

<img width="1422" alt="Screen Shot 2022-01-02 at 3 44 00 PM" src="https://user-images.githubusercontent.com/64377055/147869852-0b7b75b1-1450-46b0-a281-5e4d9b2bbec9.png">

```python
2022-01-02 15:43:34,090 - 14453 - hummingbot.connector.exchange.binance.binance_exchange - NETWORK - Error submitting BUY LIMIT_MAKER order to Binance for 0.00030000 BTC-USDT 46722.90000000.
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot_c1/hummingbot/connector/exchange/binance/binance_exchange.py", line 777, in create_order
    order_result = await self._api_request(
  File "/Users/reiniersalazar/hummingbot_c1/hummingbot/connector/exchange/binance/binance_exchange.py", line 266, in _api_request
    raise IOError(f"Error fetching data from {url}. HTTP status is {response.status}.")
OSError: Error fetching data from https://api.binance.com/api/v3/order. HTTP status is 400.
```

Logs and config: [binance.zip](https://github.com/CoinAlpha/hummingbot/files/7798779/binance.zip)

RC-13 commented:
Adding additional issue seen after testing the PR on the current latest commit:
- Client creates an order that is below the exchange minimum requirements, resulting in failing to cancel the order.
![binance-spot-cancellation-issue](https://user-images.githubusercontent.com/57189990/148143446-84234d4c-da59-4244-9379-b0e26254b776.png)
- `HTTP status is 400` keeps coming up whenever an order price created by the bot matches the current mid-price of the market. In the current development branch, an API error code 2010 is displayed whenever this scenario happens.
![binance-spot-one-side-not-submitted1](https://user-images.githubusercontent.com/57189990/148143591-dd492332-9723-425c-abc0-06afefeca703.png)
```python
2022-01-05 00:23:19,262 - 5839 - hummingbot.connector.exchange.binance.binance_exchange - NETWORK - Error submitting SELL LIMIT_MAKER order to Binance for 83.00000000 DOGE-BUSD 0.16870000.
Traceback (most recent call last):
  File "/home/ubuntu/binance/hummingbot/connector/exchange/binance/binance_exchange.py", line 783, in create_order
    order_result = await self._api_request(
  File "/home/ubuntu/binance/hummingbot/connector/exchange/binance/binance_exchange.py", line 272, in _api_request
    raise IOError(f"Error fetching data from {url}. HTTP status is {response.status} ({response}).")
OSError: Error fetching data from https://api.binance.com/api/v3/order. HTTP status is 400 (RESTResponse(url='https://api.binance.com/api/v3/order', method=POST, status=400, headers=<CIMultiDictProxy('Content-Type': 'application/json;charset=UTF-8', 'Content-Length': '62', 'Connection': 'keep-alive', 'Date': 'Wed, 05 Jan 2022 00:23:19 GMT', 'Server': 'nginx', 'x-mbx-uuid': '427bcf10-c63a-409a-b0e8-905616b08197', 'x-mbx-used-weight': '22', 'x-mbx-used-weight-1m': '22', 'x-mbx-order-count-10s': '2', 'x-mbx-order-count-1d': '224', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains', 'X-Frame-Options': 'SAMEORIGIN', 'X-Xss-Protection': '1; mode=block', 'X-Content-Type-Options': 'nosniff', 'Content-Security-Policy': "default-src 'self'", 'X-Content-Security-Policy': "default-src 'self'", 'X-WebKit-CSP': "default-src 'self'", 'Cache-Control': 'no-cache, no-store, must-revalidate', 'Pragma': 'no-cache', 'Expires': '0', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 2a08551383b826c5272c6d3873169312.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'SIN2-P2', 'X-Amz-Cf-Id': '1SWqtU3fx9cH5f2nbTZO35Du89uLQNIRAmAhubhnOSP0lPDfCFN9eA==')>)).
```

aarmoa commented:
@RC-13 do you happen to have handy the configuration you were using when the latest issue happen?

RC-13 commented:
@aarmoa I updated the initial comment, I thought I added the logs and strategy. 
For the issue with creating an order below the exchange minimum, the fastest way to replicate is to open multiple order levels.
And for the HTTP 400 status, you need to set a tight spread.

RC-13 commented:
> Client creates an order that is below the exchange minimum requirements, resulting in failing to cancel the order.
- Confirming that this issue is now fixed.
> HTTP status is 400 keeps coming up whenever an order price created by the bot matches the current mid-price of the market. 
- Logging message now matches the log message in the development branch

New issues found while testing on the latest commit:
- The logging message for an order that fails to submit is displayed incorrectly. In this screenshot, it was the ask side that fails to get submitted but in the logs, it says it was the buy order.
![image](https://user-images.githubusercontent.com/57189990/148320784-4b3f349e-ef0a-4823-9ae8-63d116cc2f1c.png)
```python
2022-01-06 02:48:31,000 - 22266 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-BUSD) Creating 1 bid orders at (Size, Price): ['86 DOGE, 0.1574 BUSD']
2022-01-06 02:48:31,001 - 22266 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-BUSD) Creating 1 ask orders at (Size, Price): ['29 DOGE, 0.1576 BUSD']
2022-01-06 02:48:31,003 - 22266 - hummingbot.connector.exchange.binance.binance_exchange - WARNING - Buy order amount 0 is lower than the minimum order size 1.00000000. The order will not be created.
2022-01-06 02:48:31,005 - 22266 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1641437311000, "order_id": "x-XEKWYICX-SDEBD1641437311001225", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "binance"}
2022-01-06 02:48:31,005 - 22266 - hummingbot.connector.client_order_tracker - INFO - Order x-XEKWYICX-SDEBD1641437311001225 has failed. Order Update: OrderUpdate(trading_pair='DOGE-BUSD', update_timestamp=1641437311000, new_state=<OrderState.FAILED: 6>, client_order_id='x-XEKWYICX-SDEBD1641437311001225', exchange_order_id=None, trade_id=None, fill_price=None, executed_amount_base=None, executed_amount_quote=None, fee_asset=None, cumulative_fee_paid=None, trade_fee_percent=None)
2022-01-06 02:48:31,101 - 22266 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT_MAKER BUY order x-XEKWYICX-BDEBD1641437311001006 for 86.00000000 DOGE-BUSD.
2022-01-06 02:48:31,111 - 22266 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1641437311000, "type": "OrderType.LIMIT_MAKER", "trading_pair": "DOGE-BUSD", "amount": "86.00000000", "price": "0.15740000", "order_id": "x-XEKWYICX-BDEBD1641437311001006", "exchange_order_id": "549933246", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "binance"}
2022-01-06 02:48:42,263 - 22266 - hummingbot.client.hummingbot_application - INFO - stop command initiated.
2022-01-06 02:48:42,357 - 22266 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1641437322000, "order_id": "x-XEKWYICX-BDEBD1641437311001006", "exchange_order_id": "549933246", "event_name": "OrderCancelledEvent", "event_source": "binance"}
2022-01-06 02:48:42,357 - 22266 - hummingbot.connector.client_order_tracker - INFO - Successfully cancelled order x-XEKWYICX-BDEBD1641437311001006.
```
- Invalid order amount recorded in the History whenever a filled order is recreated, this trade wasn't recorded in the logs, however, was saved in the sqlite and csv file, resulting in incorrect performance.
![image](https://user-images.githubusercontent.com/57189990/148329415-18d0476a-0ff6-43d9-bacd-a5cc949d2e32.png)
[logs_binance-test.log](https://github.com/CoinAlpha/hummingbot/files/7819517/logs_binance-test.log)

aarmoa commented:
@RC-13 I could fix the issue in the log text when an order can't be created because of it small size.
Regarding the issue that causes trades with wrong amount in the history command, I could not reproduce it in my environment. I added however a couple of changes in places I thought could be causing the problem (the regular request of executed trades to the exchange using the REST API). In order to help us investigate the issue I have added three new columns in the trades' table inside the history command, to show the client order id, the exchange order id and the trade id (I want to see if the trades with wrong amount are trades generated during the client's current session, or if it is caused by trades from previous sessions provided by the exchange in the trades query)

RC-13 commented:
Confirming the following reported issues has been resolved on the latest commit:
> 1. The logging message for an order that fails to submit is displayed incorrectly.
> 2. Removing the binance_markets parameter

RC-13 commented:
@aarmoa For the issue of invalid order size getting recorded, I was able to capture the problem after stopping the strategy right after getting a trade fill then start the strategy again.
![binance-invalid-size](https://user-images.githubusercontent.com/57189990/148477265-edb0c8f6-ee57-46a7-8e09-113ebb79cbae.png)
Note: In the screenshot above the order amount is 2322 but in the db file, it is 19350.0

After looking into the logs, I was unable to find where this amount is coming from however, it seems that the issue is coming from the sqlite record. I compared the db file vs the csv file, and it turns out that the csv recorded the right amount.
affected trade Order_ID's:
- `x-XEKWYICX-BDEBD1641511308001423`
- ` x-XEKWYICX-BDEBD1641512833000958`

[binance-order-amount-issue.zip](https://github.com/CoinAlpha/hummingbot/files/7825737/binance-order-amount-issue.zip)

aarmoa commented:
@RC-13 I keep investigating the issue. Based on the data you attached in your last comment, I don't think the issue is in the client, because the amount values are store correctly in the CSV files. The error only appears in the DB, and not always.
I have created a separate branch (fix/trade_fill_amount_db_difference) changing the way we store the price and amount values in the DB for TradeFill elements. Instead of storing values as floats, they are stored as bigintegers, transforming them directly from Decimals/to Decimals in the client without passing them to float. That conversion could have been causing the issue. I have also added extra logs when storing the TradeFills, to verify the client is creating the object with the correct values. Could you please see if you can reproduce the issue in that branch? You will have to remove the .sqlite file because Sqlite does not support updating columns definitions (the DB in the filesystem has to be created again).

PtrckM commented:
> @RC-13 I keep investigating the issue. Based on the data you attached in your last comment, I don't think the issue is in the client, because the amount values are store correctly in the CSV files. The error only appears in the DB, and not always. I have created a separate branch (`fix/trade_fill_amount_db_difference`) changing the way we store the price and amount values in the DB for TradeFill elements. Instead of storing values as floats, they are stored as bigintegers, transforming them directly from Decimals/to Decimals in the client without passing them to float. That conversion could have been causing the issue. I have also added extra logs when storing the TradeFills, to verify the client is creating the object with the correct values. Could you please see if you can reproduce the issue in that branch? You will have to remove the .sqlite file because Sqlite does not support updating columns definitions (the DB in the filesystem has to be created again).

- set long term monitoring bot to this branch `fix/trade_fill_amount_db_difference`
  - emptied both folders `./logs` and `./data`
  - will continue to monitor, see if I will able to replicate it.

rxlxrxsx commented:
- Checking the logs of the long term running bot I noticed that there were orders that weren't successfully submitted but eventually got filled. These orders was also recorded in the sqlite and csv. In this example you can see that the unsuccessful order got filled after 20 seconds:

```python
2022-01-08 12:21:26,001 - 5608 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (FRONT-USDT) Creating 1 bid orders at (Size, Price): ['3 FRONT, 0.6187 USDT']
2022-01-08 12:21:26,002 - 5608 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (FRONT-USDT) Creating 1 ask orders at (Size, Price): ['5 FRONT, 0.6196 USDT']
2022-01-08 12:21:26,002 - 5608 - hummingbot.connector.exchange.binance.binance_exchange - WARNING - Buy order amount 0 is lower than the minimum order size 1.00000000. The order will not be created.
2022-01-08 12:21:26,007 - 5608 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1641644486000, "order_id": "x-XEKWYICX-BFTUT1641644486001948", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "binance"}
2022-01-08 12:21:26,007 - 5608 - hummingbot.connector.client_order_tracker - INFO - Order x-XEKWYICX-BFTUT1641644486001948 has failed. Order Update: OrderUpdate(trading_pair='FRONT-USDT', update_timestamp=1641644486000, new_state=<OrderState.FAILED: 6>, client_order_id='x-XEKWYICX-BFTUT1641644486001948', exchange_order_id=None, trade_id=None, fill_price=None, executed_amount_base=None, executed_amount_quote=None, fee_asset=None, cumulative_fee_paid=None, trade_fee_percent=None)
2022-01-08 12:21:26,007 - 5608 - hummingbot.connector.exchange.binance.binance_exchange - WARNING - Sell order amount 0 is lower than the minimum order size 1.00000000. The order will not be created.
2022-01-08 12:21:26,010 - 5608 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1641644486000, "order_id": "x-XEKWYICX-SFTUT1641644486002159", "order_type": "OrderType.LIMIT_MAKER", "event_name": "MarketOrderFailureEvent", "event_source": "binance"}
2022-01-08 12:21:26,010 - 5608 - hummingbot.connector.client_order_tracker - INFO - Order x-XEKWYICX-SFTUT1641644486002159 has failed. Order Update: OrderUpdate(trading_pair='FRONT-USDT', update_timestamp=1641644486000, new_state=<OrderState.FAILED: 6>, client_order_id='x-XEKWYICX-SFTUT1641644486002159', exchange_order_id=None, trade_id=None, fill_price=None, executed_amount_base=None, executed_amount_quote=None, fee_asset=None, cumulative_fee_paid=None, trade_fee_percent=None)
2022-01-08 12:21:46,234 - 5608 - hummingbot.connector.client_order_tracker - INFO - The BUY order x-XEKWYICX-BFTUT1641644486001948 amounting to 0.00520000/0 FRONT has been filled.
2022-01-08 12:21:46,234 - 5608 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (FRONT-USDT) Maker buy order of 0.00520000 FRONT filled. [clock=2022-01-08 12:21:46+00:00]
2022-01-08 12:21:46,234 - 5608 - hummingbot.connector.markets_recorder - INFO - 
****
OrderFilledEvent(timestamp=1641644506000, order_id='x-XEKWYICX-BFTUT1641644486001948', trading_pair='FRONT-USDT', trade_type=<TradeType.BUY: 1>, order_type=<OrderType.LIMIT_MAKER: 3>, price=Decimal('3220.01000000'), amount=Decimal('0.00520000'), trade_fee=TradeFee(percent=Decimal('0'), flat_fees=[('BNB', Decimal('0.00002751'))]), exchange_trade_id='725767376', leverage=1, position='NIL')

2022-01-08 12:21:46,238 - 5608 - hummingbot.connector.markets_recorder - INFO - 
@@@@
TradeFill(id=None, config_file_path='test_filled.yml', strategy='pure_market_making', market='binance', symbol='FRONT-USDT', base_asset='FRONT', quote_asset='USDT', timestamp=1641644506234, order_id='x-XEKWYICX-BFTUT1641644486001948', trade_type='BUY', order_type='LIMIT_MAKER', price=3220.01000000, amount=0.00520000, leverage=1, trade_fee={'percent': 0.0, 'flat_fees': [{'asset': 'BNB', 'amount': 2.751e-05}]}, exchange_trade_id=725767376, position=NIL)
```

- Getting errors after running `history --verbose` command. Will further investigate what triggers the issue as it was working fine when the test bot was left running.

<img width="1423" alt="Screen Shot 2022-01-09 at 8 34 14 AM" src="https://user-images.githubusercontent.com/64377055/148667176-1978b6c7-5e73-42c2-8a2b-a3efbd7db8c2.png">

```python
2022-01-09 02:26:01,833 - 5608 - hummingbot.client.hummingbot_application - ERROR - 'NoneType' object has no attribute 'exchange_order_id'
Traceback (most recent call last):
  File "/home/ubuntu/trade_fill_amount_db_difference/hummingbot/client/hummingbot_application.py", line 179, in _handle_command
    f(**kwargs)
  File "/home/ubuntu/trade_fill_amount_db_difference/hummingbot/client/command/history_command.py", line 57, in history
    self.list_trades(start_time)
  File "/home/ubuntu/trade_fill_amount_db_difference/hummingbot/client/command/history_command.py", line 224, in list_trades
    df: pd.DataFrame = TradeFill.to_pandas(queried_trades)
  File "/home/ubuntu/trade_fill_amount_db_difference/hummingbot/model/trade_fill.py", line 154, in to_pandas
    trade.order.exchange_order_id,
AttributeError: 'NoneType' object has no attribute 'exchange_order_id'
```

Logs, config and data files: [test_filled.zip](https://github.com/CoinAlpha/hummingbot/files/7834386/test_filled.zip)

rxlxrxsx commented:
> @RC-13 I keep investigating the issue. Based on the data you attached in your last comment, I don't think the issue is in the client, because the amount values are store correctly in the CSV files. The error only appears in the DB, and not always. I have created a separate branch (`fix/trade_fill_amount_db_difference`) changing the way we store the price and amount values in the DB for TradeFill elements. Instead of storing values as floats, they are stored as bigintegers, transforming them directly from Decimals/to Decimals in the client without passing them to float. That conversion could have been causing the issue. I have also added extra logs when storing the TradeFills, to verify the client is creating the object with the correct values. Could you please see if you can reproduce the issue in that branch?
> 
> IMPORTANT: The branch already includes code that migrates the DB automatically (changes the column types and migrates the existing data to comply with the new format). But once the DB is migrated, you won't be able to use it again with the current client version. You will have to restore the previous DB file manually.

- I was able to replicate the issue under the `fix/trade_fill_amount_db_difference` branch using @RC-13 steps to reproduce. Upon checking on `history --verbose`, I noticed that the order amount shown is 100 instead of 20.

<img width="1424" alt="Screen Shot 2022-01-09 at 11 52 33 AM" src="https://user-images.githubusercontent.com/64377055/148669433-e2f20ab9-8954-4a47-bcdc-7451acd63cec.png">

Sqlite files shows a different value as well:

<img width="508" alt="Screen Shot 2022-01-09 at 12 31 54 PM" src="https://user-images.githubusercontent.com/64377055/148669564-dd92d1ca-cb36-4908-84bc-3aa069347d97.png">

Filled order log snippet:

```python
2022-01-09 03:42:19,595 - 12916 - hummingbot.connector.client_order_tracker - INFO - The BUY order x-XEKWYICX-BFTUT1641699721002247 amounting to 20.00000000/20.00000000 FRONT has been filled.
2022-01-09 03:42:19,595 - 12916 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (FRONT-USDT) Maker buy order of 20.00000000 FRONT filled. [clock=2022-01-09 03:42:19+00:00]
2022-01-09 03:42:19,595 - 12916 - hummingbot.connector.markets_recorder - INFO - 
****
OrderFilledEvent(timestamp=1641699739000, order_id='x-XEKWYICX-BFTUT1641699721002247', trading_pair='FRONT-USDT', trade_type=<TradeType.BUY: 1>, order_type=<OrderType.LIMIT_MAKER: 3>, price=Decimal('0.60970000'), amount=Decimal('20.00000000'), trade_fee=TradeFee(percent=Decimal('0'), flat_fees=[('BNB', Decimal('0.00002112'))]), exchange_trade_id='1326977', leverage=1, position='NIL')

2022-01-09 03:42:19,598 - 12916 - hummingbot.connector.markets_recorder - INFO - 
@@@@
TradeFill(id=None, config_file_path='test_filled.yml', strategy='pure_market_making', market='binance', symbol='FRONT-USDT', base_asset='FRONT', quote_asset='USDT', timestamp=1641699739595, order_id='x-XEKWYICX-BFTUT1641699721002247', trade_type='BUY', order_type='LIMIT_MAKER', price=0.60970000, amount=20.00000000, leverage=1, trade_fee={'percent': 0.0, 'flat_fees': [{'asset': 'BNB', 'amount': 2.112e-05}]}, exchange_trade_id=1326977, position=NIL)
```

```python
2022-01-09 03:44:26,763 - 12916 - hummingbot.connector.client_order_tracker - INFO - The BUY order x-XEKWYICX-BFTUT1641699721002247 amounting to 40.00000000/20.00000000 FRONT has been filled.
2022-01-09 03:44:26,763 - 12916 - hummingbot.connector.markets_recorder - INFO - 
****
OrderFilledEvent(timestamp=1641699866000, order_id='x-XEKWYICX-BFTUT1641699721002247', trading_pair='FRONT-USDT', trade_type=<TradeType.BUY: 1>, order_type=<OrderType.LIMIT_MAKER: 3>, price=Decimal('0.60970000'), amount=Decimal('20.00000000'), trade_fee=TradeFee(percent=Decimal('0'), flat_fees=[('BNB', Decimal('0.00002112'))]), exchange_trade_id='1326977', leverage=1, position='NIL')

2022-01-09 03:44:26,765 - 12916 - hummingbot.connector.markets_recorder - INFO - 
@@@@
TradeFill(id=None, config_file_path='test_filled.yml', strategy='pure_market_making', market='binance', symbol='FRONT-USDT', base_asset='FRONT', quote_asset='USDT', timestamp=1641699866763, order_id='x-XEKWYICX-BFTUT1641699721002247', trade_type='BUY', order_type='LIMIT_MAKER', price=0.60970000, amount=20.00000000, leverage=1, trade_fee={'percent': 0.0, 'flat_fees': [{'asset': 'BNB', 'amount': 2.112e-05}]}, exchange_trade_id=1326977, position=NIL)
```

Logs, config and data files: [filled_trade_issue.zip](https://github.com/CoinAlpha/hummingbot/files/7834463/filled_trade_issue.zip)

RC-13 commented:
Adding some additional trade record with incorrect value from the `fix/trade_fill_amount_db_difference` branch
Some of the affected order_IDs:
x-XEKWYICX-BFTUT1641647784001266
x-XEKWYICX-BFTUT1641677037001432
x-XEKWYICX-BFTUT1641677436002689
x-XEKWYICX-BFTUT1641699721002247
Orders that are failed to be submitted but recorded as filled:
![image](https://user-images.githubusercontent.com/57189990/148707676-c12283ac-2e9a-41a6-9935-59b488d1c1ec.png)
An order that exceeded the specified order amount (20):
![image](https://user-images.githubusercontent.com/57189990/148707744-35dcf6ae-ea9a-4600-b9f6-55c8a73464b8.png)

RC-13 commented:
Adding update based on the discussion on the client standup (Asia time): Verify whether the current reported issue exists in the development branch and if it affects other connectors.
- Run a test bot on multiple exchanges (Binance, Coinbase_pro, Crypto_com, and FTX) for 2 hours and check the trade records, currently, no abnormal trade record has been spotted. We'll check back the test bots after 24 hours.

RC-13 commented:
@aarmoa The issue with conflicting order_ID/trade_ID still persisting in the latest commit. @PtrckM and I are running on two different markets and both our test bot was able to record each of our trades.
Example order_ID: `x-XEKWYICX-SDEBD1641871430001778`  and the client_order_ID: `x-XEKWYICX-SXPUT1641871441001486` in the DOGE/BUSD market, 
```python
2022-01-11 03:24:19,663 - 9054 - hummingbot.connector.client_order_tracker - INFO - Order x-XEKWYICX-SDEBD1641871430001778 has failed. Order Update: TradeUpdate(trade_id='405657291', client_order_id='x-XEKWYICX-SXPUT1641871441001486', exchange_order_id='3873818289', trading_pair='DOGE-BUSD', fill_timestamp=1641871459607, fill_price=Decimal('0.74390000'), fill_base_amount=Decimal('15.00000000'), fill_quote_amount=Decimal('11.1585000000000000'), fee_asset='BNB', fee_paid=Decimal('0.00001948'), trade_fee_percent=None)
```
on Patrick's log, the client_order_ID found on the DOGE-BUSD market is the current order_ID of the XRP-USDT order
```python
2022-01-11 03:24:19,659 - 7851 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (XRP-USDT) Maker sell order x-XEKWYICX-SXPUT1641871441001486 (15.00000000 XRP @ 0.74390000 USDT) has been completely filled. [clock=2022-01-11 03:24:19+00:00]
```
This resulted in an error with displaying the history in verbose mode
![image](https://user-images.githubusercontent.com/57189990/148879187-cb3386ca-4b5a-401d-99bb-eb01418d044b.png)
[trades_binance-test.zip](https://github.com/CoinAlpha/hummingbot/files/7843592/trades_binance-test.zip)

[Patrick-XRP-USDT.zip](https://github.com/CoinAlpha/hummingbot/files/7843594/Patrick-XRP-USDT.zip)

RC-13 commented:
Getting the following error for a specific market on binance_perpetual connector, it occurs upon starting the client.
![image](https://user-images.githubusercontent.com/57189990/149045823-023e12ad-5d9f-4e36-b8e6-b0ccd3185dfd.png)
The logs doesn't provide much details
```python
2022-01-12 00:19:02,059 - 15765 - hummingbot.connector.derivative.binance_perpetual.binance_perpetual_api_order_book_data_source - ERROR - There was an error requesting exchange info (TRB-USDT)
```

RC-13 commented:
Confirming that the two reported issues on Binance perpetual connector have been addressed in the latest commit.

rxlxrxsx commented:
- The bot was able to receive the funding payment however it showed several funding payments with different amounts.

<img width="1423" alt="Screen Shot 2022-01-12 at 4 00 44 PM" src="https://user-images.githubusercontent.com/64377055/149088464-fc3b0dbf-d2d1-49a5-851c-e7dce7ba16ae.png">

Logs and config: [funding.zip](https://github.com/CoinAlpha/hummingbot/files/7852766/funding.zip)

RC-13 commented:
- Reported issue regarding different funding payments is happening in the the binance perpetual feature branch, confirming that it's not cause by the changes in this PR.
- @aarmoa the issue with conflicting order_ID still persist in the latest commit, my test bot was able to capture the fill event in the BTC-USDT market.
![image](https://user-images.githubusercontent.com/57189990/149266551-b8a3fc2c-4b06-4001-9251-f39b1f7f8186.png)
[binance-test-logs.zip](https://github.com/CoinAlpha/hummingbot/files/7859963/binance-test-logs.zip)

aarmoa commented:
> * Reported issue regarding different funding payments is happening in the the binance perpetual feature branch, confirming that it's not cause by the changes in this PR.
> * @aarmoa the issue with conflicting order_ID still persist in the latest commit, my test bot was able to capture the fill event in the BTC-USDT market.
>   ![image](https://user-images.githubusercontent.com/57189990/149266551-b8a3fc2c-4b06-4001-9251-f39b1f7f8186.png)
>   [binance-test-logs.zip](https://github.com/CoinAlpha/hummingbot/files/7859963/binance-test-logs.zip)

Hi @RC-13. I am not being able to replicate this issue. Even when I receive updates from orders created in other clients, my client never considers them as its own updates.
Would it be possible for you include the host name and the process id of the instances that are having trade fill collisions?

RC-13 commented:
Currently getting an `unexpected clock tick error` when starting a strategy on the latest commit.
![image](https://user-images.githubusercontent.com/57189990/149424292-2f8cf6a7-3ac9-4d28-8b56-ab5a2ee6d197.png)

```python
2022-01-13 22:54:08,001 - 1429 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 719, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_tick
    self.c_apply_budget_constraint(proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making.pyx", line 897, in hummingbot.strategy.pure_market_making.pure_market_making.PureMarketMakingStrategy.c_apply_budget_constraint
    buy_fee = market.c_get_fee(self.base_asset, self.quote_asset, OrderType.LIMIT, TradeType.BUY,
  File "hummingbot/connector/exchange_base.pyx", line 74, in hummingbot.connector.exchange_base.ExchangeBase.c_get_fee
    return self.get_fee(base_currency, quote_currency, order_type, order_side, amount, price, is_maker)
  File "/home/ubuntu/binance1/hummingbot/connector/exchange/binance/binance_exchange.py", line 378, in get_fee
    raise DeprecationWarning(
DeprecationWarning: The 'get_fee' method is deprecated, use 'build_trade_fee' and 'build_perpetual_trade_fee' instead.
```
[logs_binance-pr-test.log](https://github.com/CoinAlpha/hummingbot/files/7866583/logs_binance-pr-test.log)

RC-13 commented:
> Currently getting an unexpected clock tick error when starting a strategy on the latest commit.

- Confirming that the latest commit has addressed the reported issue. Now running a test bot for a long duration to check if there will be order_ID conflicts.

petioptrv reviewed (changes requested):
Some imports are not sorted `isort -m=3 -l=120 hummingbot/connector/exchange/binance`

aarmoa commented on hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py:
```diff
@@ -145,103 +228,72 @@ async def get_new_order_book(self, trading_pair: str) -> OrderBook:
+                        snapshot: Dict[str, Any] = await self.get_snapshot(trading_pair=trading_pair)
                         snapshot_timestamp: float = time.time()
                         snapshot_msg: OrderBookMessage = BinanceOrderBook.snapshot_message_from_exchange(
                             snapshot,
```
You are right @petioptrv. And in fact we don't really need to execute the request o'clock. We just want to wait one hour between requests. I have updated the logic following that consideration.

aarmoa commented on hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py:
```diff
@@ -1,140 +1,223 @@
     TRADE_STREAM_ID = 1
     DIFF_STREAM_ID = 2
+    DIFF_EVENT_TYPE = "depthUpdate"
+    TRADE_EVENT_TYPE = "trade"
```
I moved them

aarmoa commented on hummingbot/connector/exchange/binance/binance_exchange.py:
```diff
@@ -0,0 +1,1054 @@
+        :param order_type: the type of order to create (MARKET, LIMIT, LIMIT_MAKER)
+        :param price: the order price
+        """
+        return await self.create_order(TradeType.BUY, order_id, trading_pair, amount, order_type, price)
```
Done

aarmoa commented on hummingbot/connector/exchange/binance/binance_exchange.py:
```diff
@@ -0,0 +1,1054 @@
+        failed_cancellations = [CancellationResult(oid, False) for oid in order_id_set]
+        return successful_cancellations + failed_cancellations
+
+    async def execute_cancel(self, trading_pair: str, order_id: str):
```
Done

aarmoa commented on hummingbot/connector/time_synchronizer.py:
```diff
@@ -0,0 +1,73 @@
+
+class TimeSynchronizer:
+    """
+    Used to monkey patch Binance client's time module to adjust request timestamp when needed
```
Very nice description. I took it.

aarmoa commented on hummingbot/connector/time_synchronizer.py:
```diff
@@ -0,0 +1,73 @@
+    def __init__(self, check_interval: float = 60.0):
+        self._time_offset_ms: Deque[float] = deque(maxlen=5)
+        self._set_server_time_offset_task: Optional[asyncio.Task] = None
+        self._last_update_local_time: float = self.NaN
```
Good catch. I completed the clean up.

aarmoa commented:
@CoinAlpha/qa-team the latest commits do not apply any logic change. No test required for them.

PtrckM commented:
> @CoinAlpha/qa-team the latest commits do not apply any logic change. No test required for them.

Update: updated the testbot running to latest commits now for long duration to check if there will be any issue arise.

PtrckM reviewed (approved):
Tested and approved by QA team.
Test summary:

- Confirmed reported issues are fixed and working as expected.
- Confirmed removed commands.
- Created a strategy and went through initial prompts successfully
- Confirmed wallet and exchange balances are reflecting correct
- Started the strategy without issues, confirmed orders are created and cancelled based on config
- Ran a test bot for more than 24 hours without errors
- Confirmed in logs that is no error showing when orders are filled
- Trade information are correct (price, size) based on history command output
- Compared trades in client, sqlite data and csv thru exchange all matched.

petioptrv reviewed (approved):
Changes LGTM!

dennisocana commented:
Recreated the PR in Foundation so we can start the staging process.
